### ISSUE TITLE
classCastException: CompositeTracker hibernate 6 & mapstruct

### ISSUE THREAD

**Issue by Serkan80** [quarkusio/quarkus#34603]
### Describe the bug

Hi,

I'm migrating from Quarkus 2.16.7 to Quarkus 3.2.0 and I'm using MapStruct 1.5.5 and from the migration guide of Quarkus I read that I've to change the componentModel from CDI to JAKARTA and this is somehow not working.

I get the following error:
`java.lang.ClassCastException: Object of type 'class Address' can't be cast to CompositeTracker`.

And here is my mapping:

```java

import org.mapstruct.Mapper;
import org.mapstruct.MappingTarget;

import java.time.OffsetDateTime;

import static io.quarkus.runtime.util.StringUtil.isNullOrEmpty;
import static org.mapstruct.MappingConstants.ComponentModel.JAKARTA;

@Mapper(componentModel = JAKARTA)
public interface MyMapper {

    MyDto toDto(MyEntity entity);

    MyEntity toEntity(MyDto request);

    void updateEntity(MyDto dto, @MappingTarget MyEntity entity);

    default OffsetDateTime toOffsetDateTime(String date) {
        return isNullOrEmpty(date) ? null : OffsetDateTime.parse(date, FORMATTER);
    }
}

```

And the mapper that gets generated looks fine imho:

```java

import jakarta.inject.Named;
import jakarta.inject.Singleton;
import javax.annotation.processing.Generated;  --> shouldn't this also start with jakarta ??

@Generated(
    value = "org.mapstruct.ap.MappingProcessor",
    date = "2023-07-07T11:43:15+0200",
    comments = "version: 1.5.5.Final, compiler: javac, environment: Java 17.0.5 (Eclipse Adoptium)"
)
@Singleton
@Named
public class MyMapperImpl implements MyMapper {
 ...// nothing fancy here
     public MyEntity toEntity(MyDto dto) { 
         var result = new MyEntity();
         ...
         result.legalAddress = dto.legalAddress();  ==> I get the exception here
         return result;
     }
}
```

Address object is generated with the openapi Maven plugin and looks something like this:

```java 
import com.fasterxml.jackson.annotation.JsonProperty;
import io.swagger.annotations.ApiModelProperty;
import java.util.Objects;

public class Address {
    private String city;
    private String country;
    private String region;
    private String postalCode;
    private String firstAddressLine;
    private String additionalAddressLine1;
    private String additionalAddressLine2;
    private String additionalAddressLine3;

    public Address() {
    }

    @ApiModelProperty("")
    @JsonProperty("city")
    public String getCity() {
        return this.city;
    }

    public void setCity(String city) {
        this.city = city;
    }
   
   // idem dito for the other fields

    public boolean equals(Object o) {
        if (this == o) {
            return true;
        } else if (o != null && this.getClass() == o.getClass()) {
            Address address = (Address)o;
            return Objects.equals(this.city, address.city) && Objects.equals(this.country, address.country) && Objects.equals(this.region, address.region) && Objects.equals(this.postalCode, address.postalCode) && Objects.equals(this.firstAddressLine, address.firstAddressLine) && Objects.equals(this.additionalAddressLine1, address.additionalAddressLine1) && Objects.equals(this.additionalAddressLine2, address.additionalAddressLine2) && Objects.equals(this.additionalAddressLine3, address.additionalAddressLine3);
        } else {
            return false;
        }
    }

    public int hashCode() {
        return Objects.hash(new Object[]{this.city, this.country, this.region, this.postalCode, this.firstAddressLine, this.additionalAddressLine1, this.additionalAddressLine2, this.additionalAddressLine3});
    }
}
```

And the `Address` object is an embedded DTO used in `MyEntity` as well as in `MyDto`: 

```java
@Entity
public  MyEntity extends PanacheEntityBase {
...
    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "city", column = @Column(name = "legaladdress_city")),
            @AttributeOverride(name = "country", column = @Column(name = "legaladdress_country")),
            @AttributeOverride(name = "region", column = @Column(name = "legaladdress_region")),
            @AttributeOverride(name = "postalCode", column = @Column(name = "legaladdress_postal_code")),
            @AttributeOverride(name = "firstAddressLine", column = @Column(name = "legaladdress_first_addressline")),
            @AttributeOverride(name = "additionalAddressLine1", column = @Column(name = "legaladdress_additional_addressline1")),
            @AttributeOverride(name = "additionalAddressLine2", column = @Column(name = "legaladdress_additional_addressline2")),
            @AttributeOverride(name = "additionalAddressLine3", column = @Column(name = "legaladdress_additional_addressline3")),
    })
    public Address legalAddress;
}
```

```java
public record MyDto(
        ...
        Address legalAddress
)
```
And this used to work in Quarkus 2.16.7, and `CompositeTracker` seems to be Hibernate 6 related.

And the mapper works fine in the unit test, but when I run my Integration Test (with the application started), then it fails in the mapper on the line where the address is mapped (see above).

I think it is related to Hibernate 6, that it proxies the entity and its fields.

### Output of `java -version`

jdk 17

### Quarkus version or git rev

3.2.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

maven

**Comment by Serkan80** [comment]
I found out what was causing this problem.

I was using a generated object as an embeddable, and its class doesn't have the `@Embeddable` on it. 

Now I've created a new object and added `@Embeddable` and it's working again.

But is there a way in Hibernate to use external objects as embeddables ?

**Comment by gtoison** [comment]
I ran into a very similar problem and created an issue in Hibernate's JIRA: https://hibernate.atlassian.net/browse/HHH-16952

**Comment by yrodiere** [comment]
Indeed this seems to be a bug upstream. Thanks for reporting. I'll close this, please see https://hibernate.atlassian.net/browse/HHH-16952 for further updates.

**Comment by beikov** [comment]
Note that it is necessary to annotate the embeddable class with `@Embeddable`, as otherwise there is no way for Hibernate to understand that the class must be enhanced. I will try to disable the enhancement expectation if the annotation is missing on the embeddable class as part of [HHH-16852](https://hibernate.atlassian.net/browse/HHH-16952), but just know that without enhancement, certain optimizations simply won't work.

**Comment by Sanne** [comment]
good point @beikov - now, bearing in mind that Quarkus automatically detects which classes should get enhanced (we normally do this by looking for entities with specific marker annotations), do you think we should extend this to automatically include the types being referenced via `@Embedded` ?

**Comment by yrodiere** [comment]
> do you think we should extend this to automatically include the types being referenced via `@Embedded` ?

We already do, @Sanne. See below if you want to check.

This doesn't help though, because we'll tell Hibernate ORM to enhance those classes, but Hibernate ORM won't be able to tell that those classes are embedded.

I suppose we could add the `@Embeddable` annotation through bytecode manipulation, but that seems a bit bold... I'm afraid of unintended consequences.

--- 

See:

https://github.com/quarkusio/quarkus/blob/2bc7d57fe131fc8ce4d41eed34bc0786063b923e/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java#L311-L338

https://github.com/quarkusio/quarkus/blob/2bc7d57fe131fc8ce4d41eed34bc0786063b923e/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java#L427-L459

https://github.com/quarkusio/quarkus/blob/2bc7d57fe131fc8ce4d41eed34bc0786063b923e/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java#L475-L481

https://github.com/quarkusio/quarkus/blob/2bc7d57fe131fc8ce4d41eed34bc0786063b923e/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java#L107

https://github.com/quarkusio/quarkus/blob/2bc7d57fe131fc8ce4d41eed34bc0786063b923e/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/JpaJandexScavenger.java#L141-L142

https://github.com/quarkusio/quarkus/blob/ad94afbdc9951f7c0dcd31aa7756be49223f3df8/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L1247-L1254

**Comment by beikov** [comment]
See https://github.com/hibernate/hibernate-orm/pull/7134. I could try to register the type of a field that is annotated with `@Embedded` in the `ByteBuddyEnhancementContext`, so that subsequent calls to `isCompositeClass` can use that information, but I'm not sure what the order of enhancement is i.e. if it is guaranteed that fields of composite owners are always processed first.

Wdyt, what should we do here?

**Comment by yrodiere** [comment]
> I'm not sure what the order of enhancement is i.e. if it is guaranteed that fields of composite owners are always processed first

I very much doubt you have that kind of guarantee.

> Wdyt, what should we do here?

I'd suggest adding a pre-enhancement phase that simply goes through every class to enhance, finds `@Embedded`-annotated fields and stores de corresponding types in the `ByteBuddyEnhancementContext` for use during actual enhancement?

**Comment by beikov** [comment]
>I'd suggest adding a pre-enhancement phase that simply goes through every class to enhance, finds @Embedded-annotated fields and stores de corresponding types in the ByteBuddyEnhancementContext for use during actual enhancement?

Unfortunately, I don't think this can be reasonably done, because the `jakarta.persistence.spi.ClassTransformer` contract (which is the JPA spec wrapper for this) will be called by the Jakarta EE Server only once per class and in an undefined order.
Also, if I read all the bytecode of all class files in a first run, this will slow down the whole process of bytecode enhancement.

IMO it is a reasonable limitation that we need local information like the `@Embeddable` annotation to be able to enhance a class file. Wdyt @Sanne, @sebersole?

**Comment by yrodiere** [comment]
> Unfortunately, I don't think this can be reasonably done, because the jakarta.persistence.spi.ClassTransformer contract (which is the JPA spec wrapper for this) will be called by the Jakarta EE Server only once per class and in an undefined order.

This is only for runtime enhancement though.

My proposal would work for build-time enhancement, which is what Quarkus does.

> IMO it is a reasonable limitation that we need local information like the @Embeddable annotation to be able to enhance a class file

Agreed, th elimitation is reasonable, I just don't think it's reasonable to expect people to be aware of it, given it's fairly obscure.
When bytecode enhancement is enabled, I think it would help if ORM threw an exception during mapping for any `@Embedded` applied to a non-`@Embeddable` class, because we know something will go wrong. But maybe that's what you intend to do in your PR?

**Comment by Sanne** [comment]
Right, either we automate it for Quarkus only - which has a complete view of the entities, so that should be doable as it bypasses SPI restrictions imposed by Jakarta specs - or we make sure an actionable error message guides the user.

Either approach is totally fine; the actionable error seems more sensible as it would help all users, not just the Quarkus ones.

**Comment by Serkan80** [comment]
Some questions:

- how did this work in Hibernate 5.x ? And what made it brake in Hibernate 6 ? 
- is it currently possible to use external/generated objects as Embeddables ? If yes, how ?

**Comment by beikov** [comment]
This is now fixed in 6.3: https://github.com/hibernate/hibernate-orm/pull/7134
For 6.2, we will simply disable the enhancement expectation if the annotation is missing on the embeddable class: https://github.com/hibernate/hibernate-orm/pull/7245

**Comment by mklueh** [comment]
@beikov I'm running into this issue again https://github.com/quarkusio/quarkus/issues/47199 with Hibernate 6.6.11 with all the neccessary annotations in place

**Comment by beikov** [comment]
Let's discuss on https://github.com/quarkusio/quarkus/issues/47199 @mklueh

**Comment by matejbizjak** [comment]
I am also running into this same issue with Hibernate 6.5.3.Final with all the necessary annotations in place. It works on 6.4.10.Final though.

**Comment by yrodiere** [comment]
> I am also running into this same issue with Hibernate 6.5.3.Final with all the necessary annotations in place. It works on 6.4.10.Final though.

Please open a new issue with a reproducer.

Ideally, since this is likely unrelated to Quarkus, report to https://hibernate.atlassian.net/browse/HHH and base your reproducer on https://github.com/hibernate/hibernate-test-case-templates/blob/main/orm/hibernate-orm-7/src/test/java/org/hibernate/bugs/QuarkusLikeORMUnitTestCase.java

Thank you!

EDIT: Please test on the latest Hibernate ORM 6.6 though -- 6.5 is no longer maintained.

**Comment by nyyrro** [comment]
I have the same problem with Hibernate 6.6.55.Final without using Quarkus and with all the necessary annotations in place. Have this been fixed since please? is there a workaround?

**Comment by nyyrro** [comment]
Hello again,

In order to help others with this ugly error, a possible cause is a fake getter in a @Embeddable class that makes Hibernate not enhance the @Embeddable class: a fake getter being a getter not returning a mapped field like:

public X getX(){
     return something_else_than_a_mapped_field;
}

As per someone's suggestion on another discussion, i've tried annotating that getter with @Transient but it failed. So i have renamed the getX() to returnX() and the error disappeared!

Hope this helps others.

**Comment by beikov** [comment]
If you think this is not fixed yet @nyyrro, then please create a reproducer and if you managed to reproduce the problem, attach that to a newly created Jira ticket.
